xds/internal/xdsclient/tests/lds_watchers_test.go

```diff
@@ -919,6 +943,128 @@ func (s) TestLDSWatch_NACKError(t *testing.T) {
 	if gotErr == nil || !strings.Contains(gotErr.Error(), wantListenerNACKErr) {
 		t.Fatalf("update received with error: %v, want %q", gotErr, wantListenerNACKErr)
 	}
+
+	// Verify that the expected error is propagated to the new watcher as well.
+	lw2 := newListenerWatcher()
+	ldsCancel2 := xdsresource.WatchListener(client, ldsName, lw2)
+	defer ldsCancel2()
+	u, err = lw2.updateCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("timeout when waiting for a listener resource from the management server: %v", err)
+	}
+	gotErr = u.(listenerUpdateErrTuple).err
+	if gotErr == nil || !strings.Contains(gotErr.Error(), wantListenerNACKErr) {
+		t.Fatalf("update received with error: %v, want %q", gotErr, wantListenerNACKErr)
+	}
+}
+
+// TestLDSWatch_ResourceCaching_WithNACKError covers the case where a watch is
+// registered for a resource which is already present in the cache with an old
+// good update as well as latest NACK error. The test verifies that new watcher
+// receives both good update and error without a new resource request being
+// sent to the management server.
+func (s) TestLDSWatch_ResourceCaching_NACKError(t *testing.T) {
+	secondRequestReceived := grpcsync.NewEvent()
+
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{
+		OnStreamRequest: func(id int64, req *v3discoverypb.DiscoveryRequest) error {
+			// If the version is "2", it means that a second request has been
+			// received (after an initial request and ack). The client should
+			// not send a second request if the resource is already cached.
+			if req.GetVersionInfo() == "2" {
```

Comment:
This is making an assumption that the management server is using monotonically increasing integers starting at `1` for version numbers. There is nothing in the xDS spec that requires management servers to that I believe. I might be wrong here.

But you could do something like what we do in `TestLDSWatch_ResourceCaching` to ensure that a second request is not sent out. There will be a second ACK here though, so you would not be able to use the exact logic from there, but you can get an idea.

Reply:
yeah change to first request followed by two acks but second request should not be fired. There is no mention of restriction to incremental version for resources https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#xds-ack-nack